pkg/server/tracedumper/tracedumper.go

```
@@ -0,0 +1,121 @@
+		"total size of job trace dumps to be kept. "+
+			"Dumps are GC'ed in the order of creation time. The latest dump is "+
+			"always kept even if its size exceeds the limit.",
+		500<<20, // 500MiB
+	)
+)
+
+// TraceDumper can be used to dump a zip file containing cluster wide inflight
+// trace spans for a particular trace, to a configured dir.
+type TraceDumper struct {
+	currentTime func() time.Time
+	store       *dumpstore.DumpStore
+}
+
+// PreFilter is part of the dumpstore.Dumper interface.
+func (t *TraceDumper) PreFilter(
+	ctx context.Context, files []os.FileInfo, _ func(fileName string) error,
+) (preserved map[int]bool, err error) {
+	preserved = make(map[int]bool)
+	for i := len(files) - 1; i >= 0; i-- {
+		// Always preserve the last dump in chronological order.
+		if t.CheckOwnsFile(ctx, files[i]) {
+			preserved[i] = true
+			break
+		}
+	}
+	return
+}
+
+// CheckOwnsFile is part of the dumpstore.Dumper interface.
+func (t *TraceDumper) CheckOwnsFile(ctx context.Context, fi os.FileInfo) bool {
+	return strings.HasPrefix(fi.Name(), jobTraceDumpPrefix)
+}
+
+var _ dumpstore.Dumper = &TraceDumper{}
+
+// MaybeDump attempts to dump a trace zip of cluster wide inflight trace spans
+// with traceID, to the configured dir.
+// The file names are prefixed with the timestamp of when it was written, to
+// facilitate GC of older trace zips.
+func (t *TraceDumper) MaybeDump(
+	ctx context.Context, name string, traceID int64, ie sqlutil.InternalExecutor,
+) {
+	now := t.currentTime()
+	traceZipFile := fmt.Sprintf(
+		"%s.%s.%s",
+		jobTraceDumpPrefix,
+		now.Format(timeFormat),
+		name,
+	)
+	z := zipper.MakeInternalExecutorInflightTraceZipper(ie)
+	zipBytes, err := z.Zip(ctx, traceID)
+	if err != nil {
+		log.Errorf(ctx, "failed to collect inflight trace zip: %v", err)
+		return
+	}
+	path := t.store.GetFullPath(traceZipFile) + ".zip"
+	f, err := os.Create(path)
+	if err != nil {
+		log.Errorf(ctx, "error creating file %s for trace dump: %v", path, err)
```

Is it safe to continue here if we've failed to create the file? Perhaps this method would be more readable as a nested func that returns an error which is swallowed just once at the end.

---

Done, good idea.